squidpy/gr/_ripley.py

```diff
@@ -0,0 +1,245 @@
+"""Functions for point patterns spatial statistics."""
+from typing import List, Tuple, Union, Optional, TYPE_CHECKING
+from typing_extensions import Literal
+
+from scanpy import logging as logg
+from anndata import AnnData
+
+from numpy.random import default_rng
+from scipy.spatial import Delaunay, ConvexHull
+from sklearn.neighbors import NearestNeighbors
+from sklearn.preprocessing import LabelEncoder
+from scipy.spatial.distance import pdist
+import numpy as np
+import pandas as pd
+
+from squidpy._docs import d, inject_docs
+from squidpy.gr._utils import _save_data, _assert_spatial_basis, _assert_categorical_obs
+from squidpy._constants._constants import RipleyStat
+from squidpy._constants._pkg_constants import Key
+
+__all__ = ["ripley"]
+
+
+@d.dedent
+@inject_docs(key=Key.obsm.spatial, rp=RipleyStat)
+def ripley(
+    adata: AnnData,
+    cluster_key: str,
+    mode: Literal["F", "G", "L"],
+    spatial_key: str = Key.obsm.spatial,
+    metric: str = "euclidean",
+    n_neigh: int = 2,
+    n_simulations: int = 100,
+    n_observations: int = 1000,
+    max_dist: np.float_ = None,
+    n_steps: int = 50,
+    seed: Optional[int] = None,
+    copy: bool = False,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    r"""
+    Calculate various Ripley's statistics for point processes.
+
+    According to the `'mode'` argument, it calculates either the following Ripley's statistics:
+    `{rp.F.s!r}`, `{rp.G.s!r}` or `{rp.L.s!r}` statistics.
+
+    `{rp.F.s!r}`, `{rp.G.s!r}` are defined as:
+
+    .. math::
+
+        F(t),G(t)=P( d_{{i,j}} \le t )
+
+    Where :math:`d_{{i,j}}` represents:
+
+        - distances to a random Spatial Poisson Point Process for `'F'`.
+        - distances to any other point of the dataset for `'G'`.
+
+    `{rp.L.s!r}` we first need to compute :math:`K(t)`, which is defined as:
+
+    .. math::
+
+        K(t) = \frac{{1}}{{\lambda}} \sum_{{i \ne j}} \frac{{I(d_{{i,j}}<t)}}{{n}}
+
+    and then we apply a variance-stabilizying transformation:
+
+    .. math::
+
+        L(t) = (\frac{{K(t)}}{{\pi}})^{{1/2}}
+
+    For reference, check out
+    `Wikipedia <https://en.wikipedia.org/wiki/Spatial_descriptive_statistics#Ripley's_K_and_L_functions>`_
+    or :cite:`Baddeley2015-lm`.
+
+    Parameters
+    ----------
+    %(adata)s
+    %(cluster_key)s
+    mode
+        Which Ripley's statistic to compute.
+    %(spatial_key)s
+    metric
+        Which metric to use for computing distances.
```

Comment:
I'd add reference as it's done in sklearn for the available metrics: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn-neighbors-nearestneighbors

Reply:
added direct link, potentially we can add scikit learn as intersphinx mapping? let me know what you think

Reply:
Yes, it should be added as `sklearn=("https://scikit-learn.org/stable/", None)`.